fix configuration not being reloaded, fix "generationEnabled" not being used, fix formatting when running "Sort arb Files", add option to run sort & format when manually editing

schindler-s opened the pull request.

ZiPPSolutions commented:
When working with this package we had 2 problems in our workflow:
1. some did not sort arb files before committing them
2.  some did not have "format on save" turned on in their workspace settings

due to these both problem we had many merge conflicts and I edited this package to integrate it better into our workflow

this PR solves both problems and allows users to disable them in the settings

This is my first change on a vs code extension so if you wish any changes, feel free to contact me

lsaudon commented on package.json:
```diff
@@ -32,7 +32,8 @@
     "onLanguage:dart",
-    "workspaceContains:**/*.dart"
+    "workspaceContains:**/*.dart",
+    "onLan"
```
questions: What is the "onLan" activationEvents? It's a typo error ?

schindler-s replied:
We used these changes for some months now. I made them back then and had stashed them to apply them on the latest version. 

Because of that stash and apply I can't see when I did this change. I guess I accidently added this and will remove it. Looke like there was something like "onLanguage:xxxx" like above.

lsaudon commented:
Thanks @schindler-s for your pull request.
I'll check it out as soon as I have time.
I'm not really a specialist either.

lsaudon commented on package.json:
```diff
@@ -99,6 +100,16 @@
             "template"
           ],
           "markdownDescription": "`all` copy in each files, `template` copy only in `template-arb-file`"
+        },
+        "l10nization.sortOnSave": {
+          "type": "boolean",
+          "default": true,
+          "markdownDescription": "If enabled, automatically runs sort on save"
+        },
+        "l10nization.formatArb": {
+          "type": "boolean",
+          "default": true,
+          "markdownDescription": "If enabled, formats file on save"
```
suggestion: Put below `l10nization.arbSort`.

schindler-s commented:
thanks for the feedback, i'll be back on it the next days. Tomorrow I won't have time I guess.

schindler-s commented:
nvm, just applied your suggestions

lsaudon commented:
Hi @schindler-s,
I tried your fix. I don't see any difference with the current version and whether I enable the options or not.

Maybe the problem is only on one platform.
I tested on macOS. Are you on Linux or Windows?

Do you have an example of bad formatted or sorted file ?

Thank you in advance.

schindler-s commented:
I've been doing a lot of investigation and found some issues. I've fixed them now and need to do a code cleanup before I push another commit.

1. The configuration was initialized once when the VSCode windows was loaded but not updated after that. Changing the settings had no affect until the VSCode window was reloaded. Probably this is the reason you did not see any difference when you enabled the options or not.
=> i implemented a singleton for the configuration and added an event handler that reloads the configuration when there have been changes made that affect this extension
2. When working on the previous issue I've seen one deprecated option ```flutterPubGetEnabled``` which was used to decide wether generation should be activated or not. When going to the settings this option was not available anymore but ```generationEnabled``` was there to opt in/out from that generation. But ```generationEnabled``` wasn't used anywhere in the code so that option was buggy at the moment. I fixed it to use ```generationEnabled``` now.
3. I guess this extension started with solely the feature to extract text from dart files to arb. Later the feature "sortArb" was added.
![telegram-cloud-photo-size-2-5222128191655762950-x](https://github.com/lsaudon/l10nization/assets/54683121/927349e2-cf39-4b9a-bbe3-5d3d41aa7ec2)
When adding new l10n to arb by extracting them from dart files everything was working fine since the function "toJson" (from src/extension/toJson.ts) was used.
But when using the sortArb function that method didn't get used but instead ```JSON.stringify(Object.fromEntries(sortArb(new Map<string, unknown>(Object.entries<string>(JSON.parse(text) as string)))))``` has ben used. This code didn't have the formatting enabled.
So when you've been using the "Sort arb files" function as given in the picture above you got a one liner as a result:
![telegram-cloud-photo-size-2-5222128191655762947-y](https://github.com/lsaudon/l10nization/assets/54683121/c28fc498-0310-49aa-b686-def983e77a9f)
If you have "Format on Save" in VS Code enabled that did not cause any problems because VS Code will format it then.
![Screenshot 2024-04-05 at 17 01 34](https://github.com/lsaudon/l10nization/assets/54683121/5125ba0e-09b4-46de-9ab7-b8c5a469da02)
Nonetheless I have a colleague that has it enabled but still VS Code didn't format it when using "Sort arb files"
=> I did not realize there already have been a "toJson" function available and added another function for it that uses this optionally. I will refactor this that only one place exists to stringify the map. Should we keep the possibility to deactivate it?
I guess we can remove that option because it should always format it like it does when adding it via extracting it from the dart file.
4. There already was an option "Arb Sort" but that affected only changes done by extracting arb from text but not when you manually added new localizations to the arb file. Then you had to manually run "sort arb files" which may be forgotten by some users working on that. That is the reason I added the option to run "Sort on Save" which sorts the arb file when manually saved.
=> Probably both options can be merged since why would I sort only when working with the arb files on one of both possible ways? Would i ever like to sort only when manually saving/extracting them from the text but not sort them the other way?

I will refactor that JSON.stringify issue and do a code clean up now. Going to push new stuff after that.

schindler-s commented:
check the latest changes please

lsaudon commented:
Wow, impressive work.
Thanks for taking the time for this.

I forgot to use `generationEnabled`(not good to work alone).
Yes, sorting arbs files via a command was done later in development. Also, I had decided that vscode would be the one to format the file (if formatOnSave is enabled).

lsaudon commented:
I update the dependencies and make the version as soon as I can.

schindler-s commented:
> Wow, impressive work. Thanks for taking the time for this.
> 
> I forgot to use `generationEnabled`(not good to work alone). Yes, sorting arbs files via a command was done later in development. Also, I had decided that vscode would be the one to format the file (if formatOnSave is enabled).

Well, it would be and better if vs code formats it, but its completly breaking the format since it making it a one liner as you see in the screenshot above. I think it's better this way.

I got an email that the build did not work somewhere, any information about that?

schindler-s commented:
btw, not mentioned in the changelog @lsaudon:

fix config not being reloaded

lsaudon commented:
> Well, it would be and better if vs code formats it, but its completly breaking the format since it making it a one liner as you see in the screenshot above. I think it's better this way.
> 
> I got an email that the build did not work somewhere, any information about that?

For the build, it's just a console.log in SaveAndFormat. I fix this warning after the merge.

lsaudon commented:
> fix config not being reloaded/updated

add in last changelog
